crates/compiler/parse/src/src64.rs

```diff
@@ -0,0 +1,437 @@
+        let dest = unsafe { allocate_and_pad_with_newlines(arena, capacity)? };
+
+        // Safety: we just allocated `dest` to have len >= src.len(), and they're both u8 slices.
+        unsafe {
+            ptr::copy_nonoverlapping(
+                src.as_bytes().as_ptr(),
+                dest.as_mut_ptr() as *mut u8,
+                src_len,
+            );
+        }
+
+        Some(Self {
+            // Safety: all the bytes should now be initialized, so we can transmute away their MaybeUninit marker.
+            bytes: unsafe { mem::transmute::<&'a [MaybeUninit<u8>], &'a [u8]>(dest) },
+        })
+    }
+
+    pub fn from_file(arena: &'a Bump, path: &Path) -> Result<Self, FileErr> {
+        let mut file = File::open(path).map_err(Into::<FileErr>::into)?;
+        let file_size = file.metadata()?.len() as usize;
+
+        if file_size == 0 {
+            return Err(FileErr::FileWasEmpty);
+        }
+
+        let capacity = round_up_to_nearest_64(file_size);
+
+        // Safety: we got capacity by rounding up to the nearest 64B
+        match unsafe { allocate_and_pad_with_newlines(arena, capacity) } {
+            Some(bytes) => {
+                {
+                    // Read the contents of the file directly into `bytes`. If the file size isn't a multiple of 64,
+                    // it's no problem; we already made its capacity a multiple of 64, and the last 64B now end in newlines.
+                    //
+                    // Safety: By design, read_exact returns an error if it doesn't completely fill the buffer
+                    // it was given. However, in this case, we explicitly want it to do that, in order to preserve
+                    // our newline padding at the end. This is safe because we know that the buffer's length will
+                    // actually be strictly longer than what we're setting it to, because capacity > file_size.
+                    //
+                    // The `read` function is not an appropriate substitute for `read_exact` here, because:
+                    // - It's allowed to read fewer than the specified number of bytes, which is not what we want
+                    // - The docs - https://doc.rust-lang.org/std/io/trait.Read.html#tymethod.read - explicitly say
+                    //   that "It is your responsibility to make sure that buf is initialized before calling read.
+                    //   Calling read with an uninitialized buf (of the kind one obtains via MaybeUninit<T>) is
+                    //   not safe, and can lead to undefined behavior." That UB is exactly what we would be doing here
+                    //   if we used read on this. Using `read_exact` in this unusual way is safer than using `read`
+                    //   in a way that is explicitly documented to be UB.
+                    file.read_exact(unsafe {
+                        slice::from_raw_parts_mut(bytes.as_ptr() as _, file_size)
```

Comment:
well but this is still UB, from the docs of `from_raw_parts`

> data must point to len consecutive properly initialized values of type T.

this is not what you have here. to work around issues now, I suggest we just zero-initialize the slice

Reply:
Fixed in d35fd07eb13592522c054a70f5cee92570da51da